Draft implementation of proposed syntax changes (feedback requested)

tclose opened the pull request:
NB: This is just a draft PR at this stage, but demonstrates the new "design" syntax I had in mind. It isn't connected up to the internals of the `Task` logic so everything will be broken outside of `pydra/design/tests`, but if you have a look at those tests you should be able to see what they would look like.

## Types of changes
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Summary
Attempts to streamline the public API for defining tasks and workflow as per #692 

## Checklist
- [ ] I have added tests to cover my changes (if necessary)
- [ ] I have updated documentation (if necessary)

tclose commented:
Aside from the API changes, the main conceptual change is the delay of the workflow construction until execution time. This should bring some bonus benefits, including allowing conditional logic to be based on upstream results.

The output of the design phase for Python, shell and workflow tasks are "TaskSpec" attrs classes, which specify the inputs, the function/executable/constructor, the nested outputs class and the type of task to run it with (eg ShellCommandTask). As such, they are lightweight and can be passed around and reused at multiple places in the code without any concept of state. Each type of task (shell, python or workflow) can be defined the "canonical" data class style definition, or using shorthand function decorators or string templates.

Not that the Workflow.construct method won't be exposed to the designers or end users of the pipelines, as it will be called internally by the WorkflowTask object created when calling the workflow spec with 'MyWorkflowSpec(...)(**exec_params)'. By delaying u til this point we avoid having to distinguish between parameters used to construct the workflow and lazy "inputs" just required at runtime, ie all inputs of the pipeline are present so can be used in the conditional logic. However, this means that nested workflows need to get rebuilt for each iteration when split, so I have allowed for inputs to be marked as lazy (and therefore not able to be used in conditionals), and cache the constructed workflows by a hash of their non lazy inputs to only rebuild the same workflow as much as we need to.

satra commented on pydra/design/tests/test_shell.py:
```diff
@@ -0,0 +1,696 @@
+def test_interface_template_w_types_and_path_template_ext():
+
+    SampleInterface = shell.define(
+        "trim-png <in_image:image/png> <out|out_image:image/png>"
```
is there an in/out option for those command tools that overwrite the input?

tclose replied:
Mmm, they are a bit tricky to handle in general in the canonical form with the hash changes (so need to copied). However, we could add special syntax to handle it like

```python
SampleInterface = shell.define(
        "trim-png <modify|image:image/png>"
)
```

which could translate to a `shell.arg(name="image", copy_mode='copy', type=image.Png)` in the inputs and a `shell.out(name="image", callable=lambda image: image, type=image.Png)` in the outputs (I think that would work)

satra replied:
sounds good.

tclose replied:
have implemented this now

tclose commented:
@effigies example Nipype interface

```python

import os

import nibabel as nb
from nipype.interfaces.base import File, SimpleInterface, TraitedSpec, isdefined, traits


class MetricMathInputSpec(TraitedSpec):
    subject_id = traits.Str(desc='subject ID')
    hemisphere = traits.Enum(
        'L',
        'R',
        mandatory=True,
        desc='hemisphere',
    )
    metric = traits.Str(desc='name of metric to invert')
    metric_file = File(exists=True, mandatory=True, desc='input GIFTI file')
    operation = traits.Enum(
        'invert',
        'abs',
        'bin',
        mandatory=True,
        desc='operation to perform',
    )


class MetricMathOutputSpec(TraitedSpec):
    metric_file = File(desc='output GIFTI file')


class MetricMath(SimpleInterface):
    """Prepare GIFTI metric file for use in MSMSulc

    This interface mirrors the action of the following portion
    of FreeSurfer2CaretConvertAndRegisterNonlinear.sh::

        wb_command -set-structure ${metric_file} CORTEX_[LEFT|RIGHT]
        wb_command -metric-math "var * -1" ${metric_file} -var var ${metric_file}
        wb_command -set-map-names ${metric_file} -map 1 ${subject}_[L|R]_${metric}
        # If abs:
        wb_command -metric-math "abs(var)" ${metric_file} -var var ${metric_file}

    We do not add palette information to the output file.
    """

    input_spec = MetricMathInputSpec
    output_spec = MetricMathOutputSpec

    def _run_interface(self, runtime):
        subject, hemi, metric = self.inputs.subject_id, self.inputs.hemisphere, self.inputs.metric
        if not isdefined(subject):
            subject = 'sub-XYZ'

        img = nb.GiftiImage.from_filename(self.inputs.metric_file)
        # wb_command -set-structure
        img.meta['AnatomicalStructurePrimary'] = {'L': 'CortexLeft', 'R': 'CortexRight'}[hemi]
        darray = img.darrays[0]
        # wb_command -set-map-names
        meta = darray.meta
        meta['Name'] = f'{subject}_{hemi}_{metric}'

        datatype = darray.datatype
        if self.inputs.operation == 'abs':
            # wb_command -metric-math "abs(var)"
            data = abs(darray.data)
        elif self.inputs.operation == 'invert':
            # wb_command -metric-math "var * -1"
            data = -darray.data
        elif self.inputs.operation == 'bin':
            # wb_command -metric-math "var > 0"
            data = darray.data > 0
            datatype = 'uint8'

        darray = nb.gifti.GiftiDataArray(
            data,
            intent=darray.intent,
            datatype=datatype,
            encoding=darray.encoding,
            endian=darray.endian,
            coordsys=darray.coordsys,
            ordering=darray.ind_ord,
            meta=meta,
        )
        img.darrays[0] = darray

        out_filename = os.path.join(runtime.cwd, f'{subject}.{hemi}.{metric}.native.shape.gii')
        img.to_filename(out_filename)
        self._results['metric_file'] = out_filename
        return runtime
```

would be 

```python

import os
import nibabel as nb
from fileformats import generic
from pydra.design import python


@python.define(
    inputs={
        "operation": python.arg(allowed_values=["invert", "abs", "bin"]),
        "hemisphere": python.arg(allowed_values=["L", "R"]),
    }
)
def MetricMath(
    hemisphere: str,
    metric: str,
    metric_file: generic.File,
    operation: str,
    subject_id: str | None = None,
) -> generic.File:
    """Prepare GIFTI metric file for use in MSMSulc

    This interface mirrors the action of the following portion
    of FreeSurfer2CaretConvertAndRegisterNonlinear.sh::

        wb_command -set-structure ${metric_file} CORTEX_[LEFT|RIGHT]
        wb_command -metric-math "var * -1" ${metric_file} -var var ${metric_file}
        wb_command -set-map-names ${metric_file} -map 1 ${subject_id}_[L|R]_${metric}
        # If abs:
        wb_command -metric-math "abs(var)" ${metric_file} -var var ${metric_file}

    We do not add palette information to the output file.

    Parameters
    ----------
    subject_id : str
        subject ID
    hemisphere : str
        hemisphere
    metric : str
        name of metric to invert
    metric_file : generic.File
        input GIFTI file
    operation : str
        operation to perform

    Returns
    -------
    metric_file: generic.File
        output GIFTI file
    """
    if subject_id is None:
        subject_id = "sub-XYZ"

    img = nb.GiftiImage.from_filename(metric_file)
    # wb_command -set-structure
    img.meta["AnatomicalStructurePrimary"] = {
        "L": "CortexLeft",
        "R": "CortexRight",
    }[hemisphere]
    darray = img.darrays[0]
    # wb_command -set-map-names
    meta = darray.meta
    meta["Name"] = f"{subject_id}_{hemisphere}_{metric}"

    datatype = darray.datatype
    if operation == "abs":
        # wb_command -metric-math "abs(var)"
        data = abs(darray.data)
    elif operation == "invert":
        # wb_command -metric-math "var * -1"
        data = -darray.data
    elif operation == "bin":
        # wb_command -metric-math "var > 0"
        data = darray.data > 0
        datatype = "uint8"

    darray = nb.gifti.GiftiDataArray(
        data,
        intent=darray.intent,
        datatype=datatype,
        encoding=darray.encoding,
        endian=darray.endian,
        coordsys=darray.coordsys,
        ordering=darray.ind_ord,
        meta=meta,
    )
    img.darrays[0] = darray

    out_filename = os.path.abspath(
        f"{subject_id}.{hemisphere}.{metric}.native.shape.gii"
    )
    img.to_filename(out_filename)
    return out_filename
```

in the new syntax

tclose commented:
Alternatively, if the task-spec was to be part of a library you may want to favour the canonical dataclass-style form

```python
import os
import nibabel as nb
from fileformats import generic
from pydra.design import python


@python.define
class MetricMath(python.PythonSpec["MetricMath.Outputs"]):
    """Prepare GIFTI metric file for use in MSMSulc

    This interface mirrors the action of the following portion
    of FreeSurfer2CaretConvertAndRegisterNonlinear.sh::

        wb_command -set-structure ${metric_file} CORTEX_[LEFT|RIGHT]
        wb_command -metric-math "var * -1" ${metric_file} -var var ${metric_file}
        wb_command -set-map-names ${metric_file} -map 1 ${subject}_[L|R]_${metric}
        # If abs:
        wb_command -metric-math "abs(var)" ${metric_file} -var var ${metric_file}

    We do not add palette information to the output file.
    """

    subject_id: str | None = python.arg(help_string="subject ID", default=None)
    hemisphere = python.arg(
        allowed_values=["L", "R"],
        help_string="hemisphere",
    )
    metric: str = python.arg(help_string="name of metric to invert")
    metric_file: generic.File = python.arg(help_string="input GIFTI file")
    operation: str = python.arg(
        allowed_values=[
            "invert",
            "abs",
            "bin",
        ],
        help_string="operation to perform",
    )

    @python.outputs
    class Outputs(python.Outputs):
        metric_file: generic.File = python.out(help_string="output GIFTI file")

    @staticmethod
    def function(subject_id, hemisphere, metric, metric_file, operation):
        if subject_id is None:
            subject_id = "sub-XYZ"

        img = nb.GiftiImage.from_filename(metric_file)
        # wb_command -set-structure
        img.meta["AnatomicalStructurePrimary"] = {
            "L": "CortexLeft",
            "R": "CortexRight",
        }[hemisphere]
        darray = img.darrays[0]
        # wb_command -set-map-names
        meta = darray.meta
        meta["Name"] = f"{subject_id}_{hemisphere}_{metric}"

        datatype = darray.datatype
        if operation == "abs":
            # wb_command -metric-math "abs(var)"
            data = abs(darray.data)
        elif operation == "invert":
            # wb_command -metric-math "var * -1"
            data = -darray.data
        elif operation == "bin":
            # wb_command -metric-math "var > 0"
            data = darray.data > 0
            datatype = "uint8"

        darray = nb.gifti.GiftiDataArray(
            data,
            intent=darray.intent,
            datatype=datatype,
            encoding=darray.encoding,
            endian=darray.endian,
            coordsys=darray.coordsys,
            ordering=darray.ind_ord,
            meta=meta,
        )
        img.darrays[0] = darray

        out_filename = os.path.abspath(
            f"{subject_id}.{hemisphere}.{metric}.native.shape.gii"
        )
        img.to_filename(out_filename)
        return out_filename
```

as it is arguably less surprising and should be statically type-checkable

tclose commented:
I have added a couple more tests to `pydra/design/tests/test_workflow.py` to demonstrate nesting workflows and in-stream conditional logic

* test_nested_workflow
* test_recursively_nested_conditional_workflow

I also added the ability to specify defaults within shell-command string templates (I went with '=' in the end as it then closely follows Python's use of ':' for type annotations and '=' for defaults in function signatures and feels natural). This is demonstrated in `pydra/design/tests/test_workflow.py::test_interface_template_with_defaults`

tclose commented:
The one thing I'm not sure of after going through the tests is the best place to specify the environment. To me it feels like an execution detail and therefore should be provided to`TaskSpec.__call__(...)`. However, you could argue that defining which image a tool is to be run within should at least hinted at in the definition.

So what are peoples' thoughts about adding some kind of `env_hint` or `docker_hint` field to the `shell.define` function/decorator, which could be a docker-path to an appropriate image. Then perhaps an environment class could be passed to the `TaskSpec.__call__(env_source=DockerHub)` that could resolve these hints to environments where appropriate. 

```python
from fileformats.medimage import NiftiGzX
from pydra.design import workflow, shell
from pydra.environment import DockerHub


@shell.define(env_hint="brainlife/fsl:6.0.4")
class Bet(shell.Spec["Bet.Outputs"]):
    in_image: NiftiGzX = shell.arg()
    ...

    @shell.outputs
    class Outputs(shell.Outputs):
        out_image: NiftiGzX = shell.outarg(path_template="{in_image}_out.nii.gz")

@workflow.define
def MyWorkflow(in_image: NiftiGzX) -> NiftiGzX:
    bet = workflow.add(Bet(in_image=in_image))
    return bet.out_image


wf = MyWorkflow(in_image="/path/to/an/image.nii.gz")

result = wf(env_source=DockerHub)
```

Suggestions welcome.

effigies commented on pydra/engine/tests/test_boutiques.py:
```diff
@@ -99,24 +94,19 @@ def test_boutiques_spec_2(data_tests_dir):
-            infile=wf.lzin.infile,
-            maskfile=wf.lzin.maskfile,
+
+    def Workflow(maskfile, infile):
```
If I understand correctly?

```suggestion
    @workflow.define
    def Workflow(maskfile, infile):
```

tclose commented:
Closing in favour of #768
